### ISSUE TITLE
Sync API: Missing unique key for transaction_details entries

### ISSUE THREAD

**Issue by haukepribnow** [paypal/PayPal-REST-API-issues#196]
# Situation
In some trial-and-error, followed by a discussion with a customer support representative, I figured that the following is expected behavior: Given one single `transaction_id` as a query, the Sync API may return *multiple* `transaction_details` entries.

This can exemplary occur if money is sent to a person in a different currency - and the recipient takes a long time to actually accept the payment. Here, you may have two `transaction_details` entries for this single transaction: one with status "pending" (P) and one later one with status "successful" (S).

# Complication
While this allows precise tracking over the "transaction-internal" history (which is cool!), it does not allow having a reliable unique key for `transaction_details` entries.

This can be problematic if synching transaction information from PayPal to a custom application/database: There is no reliable way to decide if a `transaction_details` entry was already added to the database or not.

You could theoretically use the combination of `transaction_id` and `transaction_updated_date` as a key - but I don't think that there is any guarantee that two `transaction_details` entries cannot fall on the same second. If they do, this will cause the uniqueness to break - making this key unreliable.

# How to solve?
There are most likely multiple ways to resolve this. Here's my personal suggestion:

Since there has to be some PayPal-internal ID for transaction history elements, I would suggest to simply add this ID (or an anonymized version of it, if required) to each `transaction_details` entry, thereby making the entry unique.

**Comment by xiaoleih41** [comment]
The status will only goes from `pending`  to `completed`. So if they ever fall on the same second which is almost impossible. Please override the `pending` status detail with the `completed` one.

**Comment by haukepribnow** [comment]
Thank you for your reply. Unfortunately, this does not explain how to handle a situation where two partial refunds are created within the very same second.

## Example
### Query 1
Assume you run this query:
`GET /v1/reporting/transactions?page=1&page_size=1&balance_affecting_records_only=N`
(with `start_date`/`end_date` capturing the current day)

Assume you get a `transaction_detail` with `transaction_status = "V"` for some `transaction_id` and some `transaction_updated_date`.

### Query 2
Assume you then run the same query as above, just with `page = 2`.

Assume that you then get another `transaction_detail` with `transaction_status = "V"` for the same `transaction_id` and the same `transaction_updated_date`.

Now there are two possibilities on what happened. Possibility 1: There are actually **two** reversals for the single transaction. Possibility 2: A **completely new** `transaction_detail` was created that can now be found on page one.

Since the two `transaction_detail`s do not have any unique identifier, you cannot know if they are the same ones, or different but equivalent ones.

## Bottom Line

As demonstrated in this example, just "collapsing" two `transaction_detail`s with the same ID and the same update date cannot be the solution for this issue.

**Comment by xiaoleih41** [comment]
The transaction ID of the refund should be unique.  You can query the combination of refund transaction ID and transaction status instead.

**Comment by xiaoleih41** [comment]
Close this issue as the answer has been given.  Please resubmit if you have any additional questions.
